assets/src/features/devices/useUserMedia.tsx

```diff
@@ -0,0 +1,451 @@
+  saveLastAudioDevice: (info: MediaDeviceInfo) => void;
+  getLastVideoDevice: () => MediaDeviceInfo | null;
+  saveLastVideoDevice: (info: MediaDeviceInfo) => void;
+  videoTrackConstraints: boolean | MediaTrackConstraints;
+  audioTrackConstraints: boolean | MediaTrackConstraints;
+  refetchOnMount: boolean;
+};
+
+export type UseUserMediaStartConfig = {
+  audioDeviceId?: string;
+  videoDeviceId?: string;
+};
+
+export type UseUserMedia = {
+  data: UseUserMediaState | null;
+  start: (config: UseUserMediaStartConfig) => void;
+  stop: (type: MediaType) => void;
+  setEnable: (type: MediaType, value: boolean) => void;
+  init: (videoParam: boolean | MediaTrackConstraints, audioParam: boolean | MediaTrackConstraints) => void;
+};
+
+const prepareMediaTrackConstraints = (
+  deviceId: string | undefined,
+  constraints: MediaTrackConstraints | undefined
+): MediaTrackConstraints | boolean => {
+  if (!deviceId) return false;
+  const exactId: Pick<MediaTrackConstraints, "deviceId"> = deviceId ? { deviceId: { exact: deviceId } } : {};
+  return { ...constraints, ...exactId };
+};
+
+const getDeviceInfo = (trackDeviceId: string | null, devices: MediaDeviceInfo[]): MediaDeviceInfo | null =>
+  trackDeviceId ? devices.find(({ deviceId }) => trackDeviceId === deviceId) || null : null;
+
+type CurrentDevices = { videoinput: MediaDeviceInfo | null; audioinput: MediaDeviceInfo | null };
+
+const getCurrentDevicesSettings = (
+  requestedDevices: MediaStream,
+  mediaDeviceInfos: MediaDeviceInfo[]
+): CurrentDevices => {
+  const currentDevices: CurrentDevices = { videoinput: null, audioinput: null };
+
+  requestedDevices.getTracks().forEach((track) => {
+    const settings = track.getSettings();
+    if (settings.deviceId) {
+      const currentDevice = mediaDeviceInfos.find((device) => device.deviceId == settings.deviceId);
+      const kind = currentDevice?.kind || null;
+      if ((currentDevice && kind && kind === "videoinput") || kind === "audioinput") {
+        currentDevices[kind] = currentDevice || null;
+      }
+    }
+  });
+  return currentDevices;
+};
+
+const isDeviceDifferentFromLastSession = (lastDevice: MediaDeviceInfo | null, currentDevices: MediaDeviceInfo | null) =>
+  (lastDevice && currentDevices?.deviceId !== lastDevice.deviceId) || currentDevices?.label === lastDevice?.label;
+
+const isAnyDeviceDifferentFromLastSession = (
+  lastVideoDevice: MediaDeviceInfo | null,
+  lastAudioDevice: MediaDeviceInfo | null,
+  currentDevices: CurrentDevices | null
+) =>
+  isDeviceDifferentFromLastSession(lastVideoDevice, currentDevices?.videoinput || null) ||
+  isDeviceDifferentFromLastSession(lastAudioDevice, currentDevices?.audioinput || null);
+
+const stopTracks = (requestedDevices: MediaStream) => {
+  requestedDevices.getTracks().forEach((track) => {
+    track.stop();
+  });
+};
+
+const getExactConstraints = (
+  shouldAskForVideo: boolean,
+  videoConstraints: MediaTrackConstraints | undefined,
```

Comment:
It couldn't be made optional because it is not the last parameter. However, I don't want to make it optional because `undefined` is the appropriate value like `MediaTrackConstraints`

<img width="650" alt="image" src="https://github.com/membraneframework/membrane_videoroom/assets/19818474/28483dd6-79ae-4992-a31c-3daa10f9b17f">